File: Retriever.py
import sys
import time
import struct
import array

def write_ascii_values_to_file(input_string, offset):
    try:
        return array.array('i', [ord(char) - offset for char in input_string])
    except Exception as e:
        print(f"Error: {e}")


def write_bytearray_to_large_file(file_path, int_array):
    try:
        with open(file_path, 'ab') as file:
            byte_array = struct.pack('i' * len(int_array), *int_array)
            file.write(byte_array)
    except Exception as e:
        print(f"Error happen: {e}")


def retrieve(input_file, output_file_name_DNA, output_file_name_qualityScore, mode='1234'):
    try:
        count = 0
        DNA_Output = ''
        SCORE_Output = ''
        start_time = time.time()
        binFileName = output_file_name_qualityScore[:-4] + "_bin.bin"

        with open(input_file, "r") as fileobject, \
             open(output_file_name_DNA, 'a', encoding='utf-8') as dna_file, \
             open(output_file_name_qualityScore, 'a', encoding='utf-8') as score_file, \
             open(binFileName, 'ab') as bin_file:

            for line in fileobject:
                if line[0] != '@':
                    count += 1
                    if count > 5000000:
                        break
                    x = line.split()
                    DNA_Output = x[9]
                    dna_file.write(DNA_Output)
                    SCORE_Output = x[10]
                    score_file.write(SCORE_Output)
                    x = write_ascii_values_to_file(SCORE_Output, 31)
                    x.tofile(bin_file)

        assert len(DNA_Output) == len(
            SCORE_Output), f"String length unmatched, {output_file_name_DNA} length: {len(DNA_Output)}, " \
                           f"while {output_file_name_qualityScore} length: {len(SCORE_Output)}"

        end_time = time.time()
        elapsed_time = end_time - start_time
        print(f"Elapsed time: {elapsed_time} seconds")
        print(f"DNA content written to {output_file_name_DNA}")
        print(f"SCORE content written to {output_file_name_qualityScore}")
        print(f"Binary Score content written to {binFileName}")

    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python script_name.py input_file output_file_name_DNA output_file_name_qualityScore")
        sys.exit(1)

    input_file_name = sys.argv[1]
    output_file_name_DNA = sys.argv[2]
    output_file_name_qualityScore = sys.argv[3]
    retrieve(input_file_name, output_file_name_DNA, output_file_name_qualityScore)

File: Retriever.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>
#include <cstdlib>
#include <sstream>

void splitString(const std::string &str, std::string &DNA_line, std::string &SCORE_line) {
    std::stringstream ss(str);
    std::string word;
    int wordCount = 0;
    while (ss >> word && wordCount < 11) {
        if (wordCount == 9) {
            DNA_line = word;
        } else if (wordCount == 10) {
            SCORE_line = word;
        }
        ++wordCount;
    }
}

void retrieve(const std::string &inputFileName, const std::string &outputFileNameDNA,
              const std::string &outputFileNameQualityScore) {
    auto startTime = std::chrono::high_resolution_clock::now();

    std::ifstream inputFile(inputFileName);
    std::ofstream dnaFile(outputFileNameDNA);
    std::ofstream scoreFile(outputFileNameQualityScore);
    std::string binFileName = outputFileNameQualityScore.substr(0, outputFileNameQualityScore.size() - 4) + "_bin.bin";
    std::ofstream binFile(binFileName, std::ios::binary);

    if (!inputFile) {
        std::cerr << "Error opening input file!" << std::endl;
        return;
    }
    if (!dnaFile) {
        std::cerr << "Error opening DNA output file!" << std::endl;
        return;
    }
    if (!scoreFile) {
        std::cerr << "Error opening score output file!" << std::endl;
        return;
    }
    if (!binFile) {
        std::cerr << "Error opening binary output file!" << std::endl;
        return;
    }

    std::cout << "Files opened successfully." << std::endl;

    const int readBatchSize = 100000;  // Number of lines to read in each batch
    std::string line, batchDNA, batchSCORE;
    int count = 0;
    std::vector<float> buffer;


    while (std::getline(inputFile, line) && count < 5000000) {
        if (line[0] != '@') {
            std::string DNA_Output, SCORE_Output;
            splitString(line, DNA_Output, SCORE_Output);
            if (!DNA_Output.empty() && !SCORE_Output.empty()) {
                batchDNA += DNA_Output;
                batchSCORE += SCORE_Output;

                for (char ch : SCORE_Output) {
                    float f = static_cast<float>(static_cast<int>(ch) - 33);
                    buffer.push_back(f);
                }

                if (++count % readBatchSize == 0) {
                    dnaFile << batchDNA;
                    scoreFile << batchSCORE;
                    binFile.write(reinterpret_cast<const char*>(buffer.data()), buffer.size() * sizeof(float));
                    batchDNA.clear();
                    batchSCORE.clear();
                    buffer.clear();
                }
            }
        }
    }

    if (!batchDNA.empty()) {
        dnaFile << batchDNA;
        scoreFile << batchSCORE;
        binFile.write(reinterpret_cast<const char*>(buffer.data()), buffer.size() * sizeof(float));
    }

    auto elapsedTime = std::chrono::high_resolution_clock::now() - startTime;
    std::cout << "Elapsed time: "
              << std::chrono::duration_cast<std::chrono::seconds>(elapsedTime).count()
              << " seconds" << std::endl;
    std::cout << "DNA content written to " << outputFileNameDNA << std::endl;
    std::cout << "SCORE content written to " << outputFileNameQualityScore << std::endl;
    std::cout << "Binary Score content written to " << binFileName << std::endl;
}

int main(int argc, char *argv[]) {
    if (argc != 4) {
        std::cout << "Usage: ./program_name input_file output_file_name_DNA output_file_name_qualityScore" << std::endl;
        return 1;
    }
    std::string inputFileName = argv[1];
    std::string outputFileNameDNA = argv[2];
    std::string outputFileNameQualityScore = argv[3];
    retrieve(inputFileName, outputFileNameDNA, outputFileNameQualityScore);
    return 0;
}
